### PR TITLE
Set up models and services for granular provider notification settings

### PR DESCRIPTION
**Pull request by AgaDufrat**
## Context

As a provider user 
I need the ability to choose the type of email notification I receive
So that I only get notified about the things I need

## Changes proposed in this pull request

- add a feature flag
- add ProviderUserNotification model
- update NotificationsList service
- add a service to update all notification events
- update email sending services to account for different type of preferences

No UI changes.

## Guidance to review

Initially wanted to call the table 'provider_user_notification_preferences' but index name was too long.

To test:
- turn on the feature flag `FeatureFlag.activate(:configurable_provider_notifications)`
- create ProviderUserNotification  in the rails console and change it via provider interface and support interface
- check the values in the rails console

## Link to Trello card

https://trello.com/c/AtNkf165/3438-%F0%9F%8F%88-set-up-models-and-services-for-granular-provider-notification-settings

## Things to check

- [ ] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] API release notes have been updated if necessary
- [ ] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/docs/environment-variables.md#azure-hosting-devops-pipeline)

### CODE REVIEW

File: app/services/notifications_list.rb [hunk 1]
```diff
@@ -1,9 +1,20 @@
...
-  def self.for(application_choice, include_ratifying_provider: false)
-    return application_choice.provider.provider_users.where(send_notifications: true) if application_choice.accredited_provider.nil? || !include_ratifying_provider
+  attr_reader :event
 
-    application_choice.provider.provider_users.where(send_notifications: true).or(
-      application_choice.accredited_provider.provider_users.where(send_notifications: true),
-    )
+  def self.for(application_choice, include_ratifying_provider: false, event: nil)
+    if FeatureFlag.active?(:configurable_provider_notifications)
+      raise 'Undefined type of notification event' if event.nil? || !ProviderUserNotification.method_defined?(event)
```
Comment (duncanjbrown) [inline app/services/notifications_list.rb:6]: This binds us 1:1 event->notification type — is that ok? (if so, probably no need to change this for an explicit mapping at this stage)

File: app/models/provider_user_notification.rb [hunk 2]
```diff
@@ -0,0 +1,3 @@
+class ProviderUserNotification < ActiveRecord::Base
```
Comment (duncanjbrown) [inline app/models/provider_user_notification.rb:1]: Would be good to have `ProviderUserNotificationPreferences`! It reads a bit like it's a notification itself. I think you can do something like `self.table_name = :provider_user_notifications`. Or alternatively manually create the index in the migration and give it a shorter name there.

File: app/services/update_all_notification_preferences.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
+class UpdateAllNotificationPreferences
```
Comment (duncanjbrown) [inline app/services/update_all_notification_preferences.rb:1]: This is such a thin class, I'd consider adding methods straight to the model?

```ruby
class ProviderUserNotificationPreferences
  belongs_to :provider_user
  NOTIFICATION_PREFERENCES = [:application_received...].freeze

  def update_all_preferences(value)
    NOTIFICATION_PREFERENCES.each { |n| update(n => value) }
  end

  def self.notification_preference_exists?(notification_name)
    NOTIFICATION_PREFERENCES.include? notification_name
  end
end
```

File: app/models/provider_user.rb [hunk 4]
```diff
@@ -4,6 +4,7 @@ class ProviderUser < ActiveRecord::Base
   has_many :provider_permissions, dependent: :destroy
   has_many :providers, through: :provider_permissions
   has_many :notes, dependent: :destroy
+  has_one :provider_user_notification
```
Comment (duncanjbrown) [inline app/models/provider_user.rb:7]: Maybe call this association `notification_preferences`?

File: spec/services/withdraw_application_spec.rb [hunk 5]
```diff
@@ -41,6 +41,54 @@
       expect(ratifying_provider_email['rails-mail-template'].value).to eq('application_withdrawn')
     end
 
+    context 'when the configurable provider notifications feature flag is off' do
```
Comment (duncanjbrown) [inline spec/services/withdraw_application_spec.rb:44]: Can we lose the above spec as this effectively duplicates it?

**Review by duncanjbrown — COMMENTED** [review]
Looking good!

File: app/services/notifications_list.rb (hunk 1 above)
Comment (AgaDufrat) [reply app/services/notifications_list.rb:6]: I think `notification_type` may be a better name for this arg. 
For users with `application_received: true` we send up to 2 emails (application received and reminder email to the training providers), similarly for `offer_declined: true` there are 2 possible events (offer declined by the candidate and offer decline by default).

Comment (duncanjbrown) [reply app/services/notifications_list.rb:6]: Sure. Do we want to map in here or have the callers effectively do the mapping when they pass the `notification_type`?

File: spec/services/accept_offer_spec.rb [hunk 6]
```diff
@@ -27,20 +27,44 @@
...
-      application_choice = create(:application_choice, :with_offer, course_option: course_option)
+        ratifying_provider = create(:provider)
+        ratifying_provider_user = create(:provider_user, send_notifications: true, providers: [ratifying_provider])
 
-      expect { described_class.new(application_choice: application_choice).save! }.to change { ActionMailer::Base.deliveries.count }.by(3)
-      expect(ActionMailer::Base.deliveries.first.subject).to match(/has accepted your offer/)
-      expect(ActionMailer::Base.deliveries.first.to).to eq [training_provider_user.email_address]
-      expect(ActionMailer::Base.deliveries.second.to).to eq [ratifying_provider_user.email_address]
+        course_option = course_option_for_accredited_provider(provider: training_provider, accredited_provider: ratifying_provider)
+        application_choice = create(:application_choice, status: :offer, course_option: course_option)
```
Comment (duncanjbrown) [inline spec/services/accept_offer_spec.rb:41]: Why does this need to change? `with_offer` ought to be more realistic?

**Review by duncanjbrown — APPROVED** [review]
Great! One last (minor) q inline

**Review by duncanjbrown — APPROVED** [review]

**Pull request merged**